examples_tests/31.SkinningDataBenchmark/main.cpp

```diff
@@ -4,396 +4,385 @@
-        uboMgr.bind(0u, 0, 16*sizeof(float));
-
-		driver->beginScene(true, true, video::SColor(255, 0, 0, 255));
+        timeMS = std::chrono::duration_cast<std::chrono::milliseconds>(device->getTimer()->getTime()).count();
+        camera->OnAnimate(timeMS);
+        camera->render();
 
-        queries[itr] = driver->createElapsedTimeQuery();
-        driver->beginQuery(queries[itr]);
+        //construct model matrices
+        for (uint32_t i = 0; i < translationMatrices_2.size(); i++)
+        {
+            core::matrix3x4SIMD rotationMatrix;
+            rotationMatrix.setRotation(core::quaternion(0.0f, rps[i] * (timeMS / 1000.0f), 0.0f));
+            boneMatrices[i] = core::concatenateBFollowedByA(translationMatrices_2[i], rotationMatrix);
+        }
 
-		smgr->drawAll();
-        if (gpumesh)
+        if constexpr (packedMatrices)
         {
-            for (size_t i = 0u; i < gpumesh->getMeshBufferCount(); ++i)
+            driver->updateBufferRangeViaStagingBuffer(drawDataBuffer[0].get(), 0u, sizeof(core::matrix3x4SIMD) * boneMatrices.size(), boneMatrices.data());
+            driver->pushConstants(gpuPipelineLayout[0].get(), asset::ISpecializedShader::ESS_VERTEX, 0u, sizeof(core::matrix4SIMD), camera->getConcatenatedMatrix().pointer());
+        }
+        else
+        {
+            //create array of matrix columns
+            for (uint32_t i = 0u; i < translationMatrices_2.size(); i++)
             {
-                auto matCopy = gpumesh->getMeshBuffer(i)->getMaterial();
-				matCopy.MaterialType = smaterial.MaterialType;
-                if (method == 0)
-					matCopy.TextureLayer[3u].Texture = matCopy.TextureLayer[3u].Texture;
-
-                driver->setMaterial(matCopy);
-                driver->drawMeshBuffer(gpumesh->getMeshBuffer(i));
+                const auto& row0 = core::matrix4SIMD(boneMatrices[i]).getRow(0);
+                const auto& row1 = core::matrix4SIMD(boneMatrices[i]).getRow(1);
+                const auto& row2 = core::matrix4SIMD(boneMatrices[i]).getRow(2);
+                const auto& row3 = core::matrix4SIMD(boneMatrices[i]).getRow(3);
+
+                for (uint32_t j = 0u; j < 4u; j++)
+                {
+                    boneMatricesColumns[s3pc.modelMatrixOffsets[j] + i] = core::vectorSIMDf(row0[j], row1[j], row2[j], row3[j]);
+                }
+                    
             }
-        }
-        driver->endQuery(queries[itr]);
-
-		driver->endScene();
 
-#ifndef BENCH
-		std::wostringstream str;
-		str << L"Builtin Nodes Demo - Irrlicht Engine [" << driver->getName() << "] FPS:" << driver->getFPS() << " PrimitvesDrawn:" << driver->getPrimitiveCountDrawn();
+            s3pc.vp = camera->getConcatenatedMatrix();
 
-		device->setWindowCaption(str.str());
-#endif
-
-        ++itr;
-	}
+            driver->updateBufferRangeViaStagingBuffer(drawDataBuffer[1].get(), 0u, sizeof(core::vectorSIMDf)* boneMatricesColumns.size(), boneMatricesColumns.data());
```

Comment:
1. why packing of matrices happens every frame? do it once
2. make function for each packing test case of matrices that returns buffer with matrices + push constants (as a SPushConstantRange and uint8_t[128] for example) with offsets compatible with what shader expects

Reply:
> why packing of matrices happens every frame? do it once
it doesn't matter for DEBUG if its done with CPU or GPU, but we need to see some updates to see that the shader gets the correct memory locations (so benchmark is not spoiled by cache).

When actually benchmarking there should be no per-frame updates of matrices at all.